### PR TITLE
[url_launcher_web] Migrate to null-safety

### PR DESCRIPTION
**Pull request by ditman**
This PR migrates `url_launcher_web` to null-safety. It uses the "codegen" path for its Mockito mocks. Minimal changes required for testing to work (other than null-safety related syntax changes).

This *cannot* be re-endorsed yet, because it uses a more modern version of `integration_test` that is not available in `stable` yet. Once the current master rolls to stable, this could be re-endorsed.

Fixes https://github.com/flutter/flutter/issues/75244

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/url_launcher/url_launcher_web/test/pubspec.yaml [hunk 1]
```diff
@@ -14,9 +15,9 @@ dev_dependencies:
   flutter_test:
     sdk: flutter
   http: ^0.12.2
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/test/pubspec.yaml:17]: This needs to be moved to the null-safe version. Why isn't anything complaining? Unused import?

File: packages/url_launcher/url_launcher_web/test/integration_test/url_launcher_web_test.dart [hunk 2]
```diff
@@ -2,35 +2,39 @@
...
-      mockNavigator = _MockNavigator();
+      mockWindow = MockWindow();
+      mockNavigator = MockNavigator();
       when(mockWindow.navigator).thenReturn(mockNavigator);
 
+      // Simulate that window.open does something.
+      when(mockWindow.open(any, any)).thenReturn(MockWindow());
+
+      when(mockNavigator.vendor).thenReturn('Google LLC');
+      when(mockNavigator.appVersion).thenReturn('Mock version!');
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/test/integration_test/url_launcher_web_test.dart:36]: According to dart:html these can't be null, so we must mock them by default. These are used pretty much for everything the plugin does!

File: packages/url_launcher/url_launcher_web/test/build.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+targets:
+  $default:
+    sources:
+      - integration_test/*.dart
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/test/build.yaml:4]: This file is required so the build process is able to find the `@GenerateMocks` annotation

File: packages/url_launcher/url_launcher_web/test/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+integration_test/*.mocks.dart
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/test/.gitignore:1]: (Do not push generated files)

File: packages/url_launcher/url_launcher_web/pubspec.yaml [hunk 5]
```diff
@@ -4,7 +4,7 @@ homepage: https://github.com/flutter/plugins/tree/master/packages/url_launcher/u
 # 0.1.y+z is compatible with 1.0.0, if you land a breaking change bump
 # the version to 2.0.0.
 # See more details: https://github.com/flutter/flutter/wiki/Package-migration-to-1.0.0
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/pubspec.yaml:6]: Remove this comment, it's a lie now

File: packages/url_launcher/url_launcher/pubspec.yaml [hunk 6]
```diff
@@ -12,9 +12,8 @@ flutter:
         pluginClass: UrlLauncherPlugin
       ios:
         pluginClass: FLTURLLauncherPlugin
-      # TODO(mvanbeusekom): Temporary disabled until web is migrated to nnbd (advised by @blasten).
-      #web:
-      #  default_package: url_launcher_web
+      web:
+        default_package: url_launcher_web
```
Comment (ditman) [inline packages/url_launcher/url_launcher/pubspec.yaml:16]: Remove this, until `integration_test` from sdk lands on master, and all plugins can be migrated to that!

**Review by ditman — COMMENTED** [review]
Some notes/comments I left to myself. I'll do the changes later.

File: packages/url_launcher/url_launcher/pubspec.yaml (hunk 6 above)
Comment (ditman) [reply packages/url_launcher/url_launcher/pubspec.yaml:16]: (This is currently breaking CHANNEL:stable tests, as expected)

**Comment by mvanbeusekom** [comment]
@googlebot I consent.

File: packages/url_launcher/url_launcher_web/test/.gitignore (hunk 4 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/test/.gitignore:1]: This needs to be pushed, or analyzer won't pass

File: packages/url_launcher/url_launcher/pubspec.yaml (hunk 6 above)
Comment (ditman) [reply packages/url_launcher/url_launcher/pubspec.yaml]: Reverted, but STABLE is not failing because of this.

File: packages/url_launcher/url_launcher_web/test/pubspec.yaml (hunk 1 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/test/pubspec.yaml]: Cleaned up

File: packages/url_launcher/url_launcher_web/pubspec.yaml (hunk 5 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/pubspec.yaml]: Cleaned up

File: packages/url_launcher/url_launcher_web/example/integration_test/link_widget_test.dart [hunk 7]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/example/integration_test/link_widget_test.dart:3]: (This file is split from the old url_launcher_web_test)

File: packages/url_launcher/url_launcher_web/lib/src/link.dart [hunk 8]
```diff
@@ -45,16 +45,16 @@ class WebLinkDelegate extends StatefulWidget {
...
 class WebLinkDelegateState extends State<WebLinkDelegate> {
-  LinkViewController _controller;
+  late LinkViewController _controller;
 
   @override
   void didUpdateWidget(WebLinkDelegate oldWidget) {
     super.didUpdateWidget(oldWidget);
     if (widget.link.uri != oldWidget.link.uri) {
-      _controller?.setUri(widget.link.uri);
+      _controller.setUri(widget.link.uri!);
```
Comment (mdebbar) [inline packages/url_launcher/url_launcher_web/lib/src/link.dart:54]: Having a null `uri` is valid.

File: packages/url_launcher/url_launcher_web/lib/src/link.dart [hunk 9]
```diff
@@ -78,7 +78,7 @@ class WebLinkDelegateState extends State<WebLinkDelegate> {
             onCreatePlatformView: (PlatformViewCreationParams params) {
               _controller = LinkViewController.fromParams(params, context);
               return _controller
-                ..setUri(widget.link.uri)
+                ..setUri(widget.link.uri!)
```
Comment (mdebbar) [inline packages/url_launcher/url_launcher_web/lib/src/link.dart:81]: Same

File: packages/url_launcher/url_launcher_web/lib/src/link.dart [hunk 10]
```diff
@@ -207,7 +207,7 @@ class LinkViewController extends PlatformViewController {
     pushRouteNameToFramework(context, routeName);
   }
 
-  Uri _uri;
+  late Uri _uri;
```
Comment (mdebbar) [inline packages/url_launcher/url_launcher_web/lib/src/link.dart:210]: We should accept null values here.

**Review by mdebbar — CHANGES_REQUESTED** [review]
Thanks for doing the migration! 🎉

File: packages/url_launcher/url_launcher_web/lib/src/link.dart (hunk 10 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/lib/src/link.dart:210]: Good to know! Thanks for pointing this out, I'll fix it ASAP!

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml [hunk 11]
```diff
@@ -1,22 +1,22 @@
...
     sdk: flutter
-  http: ^0.12.2
-  mockito: ^4.1.1
-  url_launcher_web:
-    path: ../
   integration_test:
-    path: ../../../integration_test
+    sdk: flutter
 
+environment:
```
Comment (kevmoo) [inline packages/url_launcher/url_launcher_web/example/pubspec.yaml:20]: we generally leave this at the top – why the change?

File: packages/url_launcher/url_launcher_web/pubspec.yaml [hunk 12]
```diff
@@ -14,31 +11,18 @@ flutter:
...
 dependencies:
-  url_launcher_platform_interface: ^1.0.9
-  # TODO(mvanbeusekom): Update to use pub.dev once null safety version is published.
-  # url_launcher_platform_interface:
-  #   git:
-  #     url: https://github.com/flutter/plugins.git
-  #     ref: nnbd
-  #     path: packages/url_launcher/url_launcher_platform_interface
+  url_launcher_platform_interface: ^2.0.0-nullsafety
+  meta: ^1.3.0-nullsafety.3
```
Comment (kevmoo) [inline packages/url_launcher/url_launcher_web/pubspec.yaml:15]: You could/should drop the `-null-safety.3` bit here! this is stable now

File: packages/url_launcher/url_launcher_web/lib/src/link.dart (hunk 10 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/lib/src/link.dart]: I've made this `Uri?`, and also added a small test to ensure null Uris are handled properly!

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml (hunk 11 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:20]: WAT? I've always seen this at the bottom of every pubspec I've touched so far!

(It looks weird at the top to me! 😅)

File: packages/url_launcher/url_launcher_web/lib/src/link.dart [hunk 13]
```diff
@@ -193,7 +193,7 @@ class LinkViewController extends PlatformViewController {
       return;
     }
 
-    if (_uri.hasScheme) {
+    if (_uri != null && _uri!.hasScheme) {
```
Comment (mdebbar) [inline packages/url_launcher/url_launcher_web/lib/src/link.dart:196]: This is a future bug caught by null safety!

**Review by mdebbar — APPROVED** [review]

File: packages/url_launcher/url_launcher_web/pubspec.yaml (hunk 12 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/pubspec.yaml:15]: Nice! I'm going to give this a shot, I hope this doesn't conflict with other plugins in the repo :P (I think I copied that dep from other plugins just to make sure it was the same!)

Comment (ditman) [reply packages/url_launcher/url_launcher_web/pubspec.yaml]: I've also noticed that pedantic 1.10.0 already has a non -nullsafe version published! Yay!

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml [hunk 14]
```diff
@@ -1,22 +1,22 @@
...
     sdk: flutter
   flutter_test:
     sdk: flutter
-  http: ^0.12.2
-  mockito: ^4.1.1
-  url_launcher_web:
-    path: ../
   integration_test:
-    path: ../../../integration_test
+    sdk: flutter
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: Can you just opt out the integration test code (`// @dart=2.9`) for now as we've been doing in other packages, to avoid having plugins in the repo using different sources of `integration_test`? Then once this is on stable we can migrate everything.

File: packages/url_launcher/url_launcher_web/example/run_test [hunk 15]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/bash
+if pgrep -lf chromedriver > /dev/null; then
```
Comment (blasten) [inline packages/url_launcher/url_launcher_web/example/run_test:2]: nit: maybe specify how to use this script here?  Also, what about adding the extension `.sh`?

File: analysis_options.yaml [hunk 16]
```diff
@@ -4,6 +4,7 @@ analyzer:
     # Ignore generated files
     - '**/*.g.dart'
     - 'lib/src/generated/*.dart'
+    - '**/*.mocks.dart' # Mockito @GenerateMocks
```
Comment (blasten) [inline analysis_options.yaml:7]: nice 👏

**Review by blasten — APPROVED** [review]

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml (hunk 14 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: If I opt out the integration test code, this package has no tests :)

File: packages/url_launcher/url_launcher_web/example/run_test (hunk 15 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/example/run_test:2]: I'm going to send people to the README.md if it fails the chromedriver check :P

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml (hunk 14 above)
Comment (natebosch) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: > to avoid having plugins in the repo using different sources of `integration_test`?

What problems arise when different plugins are using difference sources of that package?

Comment (ditman) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: > What problems arise when different plugins are using difference sources of that package?

I would expect none, since it is a dev_dependency, but dart deps still surprise me from time to time :/

My plan was to **not endorse** web plugins until all the standard plugins have rolled to integration_test from SDK, just to make sure that people can use them on the `stable` channel. 

(integration_test will not be cherry-picked for the upcoming stable, it'll probably come in the next one.)

Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: > > to avoid having plugins in the repo using different sources of `integration_test`?
> 
> What problems arise when different plugins are using difference sources of that package?

This comment wasn't prompted by a specific problem, just a preference for being able to easily reason about how `intgration_test` is working across this repo, with an all-at-once migration once it's fully ready.

Comment (ditman) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: All right, here's the latest info about this. For now, `integration_test` from `sdk` is only used to run the integration tests of the `example` app of the package (with its own separate pubspec.yaml).

I've downgraded the SDK requirement of the *core* package, so it works in the upcoming `stable`.

The package will work in `stable` and up as normal (it'll even be endorsable). HOWEVER, the tests of the package (for now) will require `master` to run (or whatever channel contains the new null-safe integration_test package in the SDK).

Since Web tests need to be run manually (and they have never run automatically in flutter/plugins CI), I think this is an acceptable tradeoff. End users will have null-safe web support out of the box, and plugin developers will have a way to write and run tests for this package.

The "disparity" will only last until integration_test from SDK becomes widely available (which is going to be in the next push to Stable in April/May?), then even people in `stable` will be able to run the integration tests inside the `example` app.

How does this sound? @stuartmorgan

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml (hunk 11 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:20]: I suspect at some point there was either a Flutter template that had it this way, or a plugin here ended up with it that way, and it's persisted/spread within the repo.

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml (hunk 14 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:18]: I think that's probably fine. Realistically, I think the set of "people who need to run flutter/plugin integration tests" is probably entirely a subset of "people who are on master". If someone really wants to run integration tests on beta, they can revert those files locally. And if we find out that it's a more common use case than we think, we can revert them here.

(I'm still trying to figure out if there's a reasonable way for us to solve this problem for all the plugins, so that we're getting NNBD integration tests on master, but still running integration tests on beta and stable. Having no NNBD integration testing makes me pretty nervous.)

File: packages/url_launcher/url_launcher_web/example/pubspec.yaml (hunk 11 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_web/example/pubspec.yaml:20]: (I would 100% approve a mass PR that moves them all to match the template, BTW, if anyone feels like doing that. I may do it at some point if not.)

**Comment by ditman** [comment]
This has been published: https://pub.dev/packages/url_launcher_web/versions/2.0.0-nullsafety
